pkg/controller/codestatus/logic.go

```diff
@@ -0,0 +1,102 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package codestatus defines a web controller for the code status page of the verification
+// server. This view allows users to view the status of previously-issued OTP codes.
+package codestatus
+
+import (
+	"fmt"
+	"net/http"
+
+	"github.com/google/exposure-notifications-verification-server/pkg/api"
+	"github.com/google/exposure-notifications-verification-server/pkg/controller"
+	"github.com/google/exposure-notifications-verification-server/pkg/database"
+)
+
+func (c *Controller) CheckCodeStatus(r *http.Request, uuid string) (*database.VerificationCode, int, *api.ErrorReturn) {
+	ctx := r.Context()
+	logger := c.logger.Named("issueapi.CheckCodeStatus")
+	authApp, user, err := c.getAuthorizationFromContext(r)
+	if err != nil {
+		return nil, http.StatusUnauthorized, api.Error(err)
+	}
+	if user == nil {
+		logger.Errorw("failed to check otp code status", "error", "no user found")
+		return nil, http.StatusUnauthorized,
+			api.Errorf("failed to check otp code status: no user found").WithCode(api.ErrVerifyCodeUserUnauth)
+	}
+
+	var realm *database.Realm
+	if authApp != nil {
+		realm, err = authApp.Realm(c.db)
+		if err != nil {
+			return nil, http.StatusUnauthorized, nil
+		}
+	} else {
+		// if it's a user logged in, we can pull realm from the context.
+		realm = controller.RealmFromContext(ctx)
+	}
+	if realm == nil {
+		return nil, http.StatusBadRequest, api.Errorf("missing realm")
+	}
+
+	code, err := c.db.FindVerificationCodeByUUID(uuid)
+	if err != nil {
+		logger.Errorw("failed to check otp code status", "error", err)
+		return nil, http.StatusInternalServerError,
+			api.Errorf("failed to check otp code status, please try again").WithCode(api.ErrInternal)
+	}
+
+	logger.Debugw("Found code", "verificationCode", code)
+
+	if code.UUID == "" { // if no row is found, code will not be populated
+		logger.Errorw("failed to check otp code status", "error", "code not found")
+		return nil, http.StatusNotFound,
+			api.Errorf("failed to check otp code status").WithCode(api.ErrVerifyCodeNotFound)
```

Comment:
we should probably add language here that indicates the code may have been cleaned up alread.